custom/ilsgateway/tanzania/reports/configs/dashboard_config.py

```diff
@@ -0,0 +1,127 @@
+from datetime import timedelta
+
+from django.db import connection
+
+from corehq.apps.locations.models import SQLLocation
+from custom.ilsgateway.models import DeliveryGroups
+from dimagi.utils.dates import get_business_day_of_month
+from dimagi.utils.decorators.memoized import memoized
+
+
+class DashboardConfig(object):
+    def __init__(self, domain, location_id, start_date, end_date):
+        self.domain = domain
+        self.location_id = location_id
+        self.start_date = start_date
+        self.end_date = end_date
+
+    @property
+    @memoized
+    def sql_location(self):
+        return SQLLocation.objects.get(domain=self.domain, location_id=self.location_id)
+
+    @property
+    @memoized
+    def descendants(self):
+        return list(self.sql_location.get_descendants().filter(location_type__administrative=False)
+                    .exclude(is_archived=True))
+
+    @property
+    @memoized
+    def case_ids(self):
+        return tuple([sql_location.supply_point_id for sql_location in self.descendants])
+
+    @memoized
+    def get_location_ids(self, group):
+        if not group:
+            return tuple([sql_location.supply_point_id for sql_location in self.descendants])
+        else:
+            return tuple(
+                [
+                    sql_location.location_id
+                    for sql_location in self.descendants
+                    if sql_location.metadata.get('group') == group
+                ]
+            )
+
+    @property
+    @memoized
+    def soh_data_total(self):
+        if not self.case_ids:
+            return {}
+        previous_month_date = (self.start_date.replace(day=1) - timedelta(days=1))
+        pm_last_business_day = get_business_day_of_month(previous_month_date.year, previous_month_date.month, -1)
+        on_time_date = get_business_day_of_month(self.start_date.year, self.start_date.month, 6)
+        cm_last_business_day = get_business_day_of_month(self.start_date.year, self.start_date.month, -1)
+        query = """
+        SELECT status, COUNT(case_id) FROM
+            (
+                SELECT DISTINCT ON (case_id) case_id,
+                CASE WHEN sr.date::date < %s THEN 'on_time' ELSE 'late' END as status
+                FROM stock_stocktransaction st JOIN stock_stockreport sr ON sr.id = st.report_id
+                WHERE sr.domain = %s AND sr.date BETWEEN %s AND %s AND case_id IN %s
+                AND st.type IN ('stockonhand', 'stockout')
+                ORDER BY case_id, sr.date
+            )x
+        GROUP BY status;
+        """
+        cursor = connection.cursor()
```

Comment:
i just realized, it would be better to use the cursor from within a context manager to ensure it is closed properly: `with connection.cursor() as cursor:`.  could you update that in a new pr?